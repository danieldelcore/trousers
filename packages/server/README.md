# @trousers/server

`@trousers/server` provides everything you need for Server side rendering (SSR) with Trousers.

SSR with Trousers follows a similar approach to [styled-components](https://www.styled-components.com/docs/advanced#server-side-rendering). It works by firstly instantiating a `serverStyleRegistry`, wrapping your application in a `ServerProvider`, then passing that registry into the provider as a prop. Then when you render your application to a string with `react-dom/server`, Trousers will push styles into the style registry. You can then pull the styles from the registry and manually append them to the head of your document.

```jsx
import React, { FC, ReactNode } from 'react';
import { renderToString } from 'react-dom/server';

import { serverRegistry, ServerProvider } from 'trousers';
import App from './';

const registry = serverRegistry();

const html = renderToString(
    <ServerProvider registry={registry}>
        <App />
    </ServerProvider>,
);

// Your styles will be accessible here
const styleTags = registry.get();
```

## API

### `serverRegistry`

A style registry **for use on the server**

**Example:**

```jsx
import { serverRegistry, ServerProvider } from 'trousers';

const registry = serverRegistry();
const styleTags = registry.get();
```

### `ServerProvider`

A context provider which tells Trousers to push styles into the supplied registry, rather than `document.head`. **For use on the server.**

**Props:**

-   `registry`: serverRegistry
-   `children`: ReactChildren

**Example:**

```jsx
import React, { FC, ReactNode } from 'react';
import { renderToString } from 'react-dom/server';

import { serverRegistry, ServerProvider } from 'trousers';
import App from './';

const registry = serverRegistry();

const html = renderToString(
    <ServerProvider registry={registry}>
        <App />
    </ServerProvider>,
);

const styleTags = registry.get();
```
