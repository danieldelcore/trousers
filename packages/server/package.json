{
    "name": "@trousers/server",
    "version": "3.0.1",
    "main": "dist/index.js",
    "module": "dist/index.js",
    "types": "dist/index.d.ts",
    "license": "MIT",
    "dependencies": {
        "@trousers/registry": "3.0.1",
        "@trousers/utils": "3.0.0"
    },
    "devDependencies": {
        "react": "^16.8.1"
    },
    "peerDependencies": {
        "react": ">=16.8.0"
    },
    "repository": "https://github.com/danieldelcore/trousers/tree/master/packages/server"
}
