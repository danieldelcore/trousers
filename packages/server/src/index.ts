export { default as serverRegistry, ServerRegistry } from './registry';
export {
    ServerCtx,
    ServerContext,
    ServerConsumer,
    ServerProvider,
    ServerProviderProps,
} from './ServerContext';
