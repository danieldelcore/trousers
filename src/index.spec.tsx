/**
 * @jest-environment node
 */

import React, { FC, ReactNode } from 'react';
import { renderToString } from 'react-dom/server';
import { hydrate } from 'react-dom';
import { JSDOM } from 'jsdom';

import {
    StyleCollector,
    SingleStyleCollector,
    ThemeProvider,
    styleCollector,
    useStyles,
    css,
    useGlobals,
    ServerProvider,
    ServerStyleRegistry,
} from './';

interface ButtonProps {
    children: ReactNode;
    primary?: boolean;
}

interface Theme {
    default: string;
    primary: string;
}

describe('Server side rendering (SSR)', () => {
    let styles: (props: ButtonProps) => StyleCollector<Theme>;
    let globalStyles: SingleStyleCollector<Theme>;
    let Button: FC<ButtonProps>;
    let theme: Theme;

    beforeEach(() => {
        theme = {
            default: '#b3cde8',
            primary: '#f95b5b',
        };

        globalStyles = css<Theme>`
            * {
                background-color: red;
            }
        `;

        styles = (props: ButtonProps) => styleCollector<Theme>('button')
            .element`
                background-color: ${theme => theme.default};
                color: white;
            `.modifier(!!props.primary)`
                background-color: ${theme => theme.primary};
                color: #ffffff;
            `;

        Button = props => {
            const classNames = useStyles(styles(props));

            return <button className={classNames}>{props.children}</button>;
        };
    });

    afterEach(() => {
        // @ts-ignore
        global.window = undefined;
        // @ts-ignore
        global.document = undefined;
    });

    it('stringify application on server and hydrate on client', () => {
        jest.spyOn(console, 'warn');

        const App: FC = () => {
            useGlobals(globalStyles);

            return (
                <ThemeProvider theme={theme}>
                    <Button>Hello, SSR!</Button>
                    <Button primary>Hello, Hydration!</Button>
                </ThemeProvider>
            );
        };

        const registry = new ServerStyleRegistry();

        const html = renderToString(
            <ServerProvider registry={registry}>
                <App />
            </ServerProvider>,
        );

        const jsdom = new JSDOM(`<!doctype html><html><body></body></html>`);
        const { window } = jsdom;

        // @ts-ignore
        global.window = window;
        // @ts-ignore
        global.document = window.document;

        const root = document.createElement('div');
        root.innerHTML = html;

        expect(() => hydrate(<App />, root)).not.toThrow();
        expect(html).toMatchSnapshot();
        expect(registry.get()).toMatchSnapshot();
        expect(console.warn).not.toHaveBeenCalled();
    });

    it('throws if ServerProvider was not used', () => {
        const App: FC = () => (
            <ThemeProvider theme={theme}>
                <Button>Hello, SSR!</Button>
                <Button primary>Hello, Hydration!</Button>
            </ThemeProvider>
        );

        expect(() => renderToString(<App />)).toThrowError(
            'Server style registry is required for SSR, did you forget to use <ServerProvider/>?',
        );
    });
});
