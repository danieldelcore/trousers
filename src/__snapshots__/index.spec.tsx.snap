// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server side rendering (SSR) stringify globals on server and hydrate on client 1`] = `<p>Hello world</p>`;

exports[`Server side rendering (SSR) stringify globals on server and hydrate on client 2`] = `
Array [
  <style
    data-trousers-global="1957078772"
  >
    *{background-color:red;}
  </style>,
  <style
    data-trousers=""
  >
    
  </style>,
]
`;

exports[`Server side rendering (SSR) stringify styles on server and hydrate on client 1`] = `
<button class="button__11079579703130397219">Hello, SSR!</button>
<button class="button__11079579703130397219 button--26780284923130397219">Hello, Hydration!</button>
`;

exports[`Server side rendering (SSR) stringify styles on server and hydrate on client 2`] = `
Array [
  <style
    data-trousers=".button__11079579703130397219"
  >
    .button__11079579703130397219{background-color:#b3cde8;color:white;}
  </style>,
]
`;

exports[`Server side rendering (SSR) stringify styles(css prop) on server and hydrate on client 1`] = `
<button class="button__11079579703130397219">Hello, SSR!</button>
<button class="button__11079579703130397219 button--26780284923130397219">Hello, Hydration!</button>
`;

exports[`Server side rendering (SSR) stringify styles(css prop) on server and hydrate on client 2`] = `
Array [
  <style
    data-trousers=".button__11079579703130397219"
  >
    .button__11079579703130397219{background-color:#b3cde8;color:white;}
  </style>,
]
`;
